### ISSUE TITLE
Event should be flushed on application shutdown

### ISSUE THREAD

**Issue by buckett** [getsentry/sentry-java#2073]
### Problem Statement

If you are using the Spring Boot starter for sentry and have a process that is quickly exiting after failing (and generating a sentry error) the errors can end up not being sent to sentry as they are still queued up.

### Solution Brainstorm

Recently support was added to synchronously flush events in #1239 and should the starter register a Spring Bean with a destroy method that flushes everything to sentry?

**Comment by bruno-garcia** [comment]
The SDK has a shutdown integration that flushes events when the JVM shuts down:

https://github.com/getsentry/sentry-java/blob/9f1af1c68c58570a731a65289183107d3c4f90d5/sentry/src/main/java/io/sentry/ShutdownHookIntegration.java#L33-L34

Turns out this is called after the Spring Boot winds down, which by convention calls `close` on any `Bean` that has a method named like that (convention). That happens to be `Hub` which closes the SDK. 

`close` should flush the events, so there's no need to add the `DisposableBean` you suggested. And we can disable `ShutdownHookIntegration` since it won't help on Spring Boot.

**Comment by adinauer** [comment]
https://github.com/getsentry/sentry-java/blob/9f1af1c68c58570a731a65289183107d3c4f90d5/sentry/src/main/java/io/sentry/SentryClient.java#L629

`close` already flushes. @buckett are you sure this is needed.

**Comment by buckett** [comment]
Ok, so the reason we weren't sending events to sentry was that we had made all the Spring Bean lazy with:

    spring.main.lazy-initialization=true

and this meant that the Sentry initialization wasn't getting run. This was clear when setting `sentry.debug=true`.

**Comment by maciejwalkowiak** [comment]
I find it interesting that with lazy initialization, the disposable bean did run on application context close. Perhaps we should include it in docs @adinauer?

**Comment by adinauer** [comment]
Hmm not sure I understand it correctly. If you enable lazy Spring Bean initialization then `Sentry.init` is never called, correct? So even with the disposable bean above Sentry would never be initialized. Should we add `@Lazy(false)` to the `IHub` bean?

**Comment by maciejwalkowiak** [comment]
I did some testing:

- with Spring MVC in place - `Hub` is created even if lazy init is on (because of servlet filters)
- without Spring MVC - `Hub` does not get created if lazy init is on

Adding `@Lazy(false)` to `IHub` bean is IMO the right way to go 👍

**Comment by buckett** [comment]
@maciejwalkowiak Yeah, we don't have Spring MVC in this particular application. We ended up just being more specific and marking specific application beans as `@Lazy` and reverting back to the default of everything not `@Lazy`.

If adding `@Lazy(false)` to `IHub` doesn't have any downsides it does seem sensible so that others don't run into this.

**Comment by buckett** [comment]
Best to open a new issue for supporting applications with lazy init or re-open this one (updating title)?

**Comment by adinauer** [comment]
I've just opened https://github.com/getsentry/sentry-java/issues/2082 to track it.
